kotlin/internal/jvm/compile.bzl

```diff
@@ -237,81 +267,174 @@ def kt_jvm_compile_action(ctx, rule_kind, output_jar, compile_jar):
-                "abi_jar": kt_compile_jar,
+                "generated_java_srcjar": kapt_generated_src_jar,
+                "kapt_generated_stub_jar": kapt_generated_stub_jar,
+                "kapt_generated_class_jar": kapt_generated_class_jar,
             },
-            mnemonic = "KotlinCompileAbi",
+            build_java = False,
+            build_kotlin = False,
+            mnemonic = "KotlinKapt",
         )
-        if not srcs.java:
-            compile_jar = kt_compile_jar
+        generated_src_jars.append(kapt_generated_src_jar)
+        output_jars.append(kapt_generated_class_jar)
+        kt_stubs_for_java.append(JavaInfo(compile_jar=kapt_generated_stub_jar, output_jar=kapt_generated_stub_jar, neverlink = True))
+
+    java_infos = []
+    # Build Kotlin
+    if srcs.kt or srcs.src_jars:
+        kt_runtime_jar = ctx.actions.declare_file(ctx.label.name + "-kt.jar")
+        if toolchains.kt.experimental_use_abi_jars:
+            kt_compile_jar = ctx.actions.declare_file(ctx.label.name + "-kt.abi.jar")
+            outputs = {
+                "output": kt_runtime_jar,
+                "abi_jar": kt_compile_jar,
+            }
         else:
-            java_compile_jar = ctx.actions.declare_file(ctx.label.name + "-java.abi.jar")
-            java_info = java_common.compile(
-                ctx,
-                source_files = srcs.java,
-                output = java_compile_jar,
-                deps = compile_deps.deps + [JavaInfo(compile_jar=kt_compile_jar, output_jar=kt_compile_jar)],
-                java_toolchain = toolchains.java,
-                host_javabase = toolchains.java_runtime,
-            )
-            compile_jar = ctx.actions.declare_file(ctx.label.name + ".abi.jar")
-            _fold_jars_action(
-                ctx,
-                rule_kind = rule_kind,
-                output_jar = compile_jar,
-                action_type = "Abi",
-                input_jars = [
-                    kt_compile_jar,
-                    java_common.run_ijar(
-                        ctx.actions,
-                        target_label = ctx.label,
-                        jar = java_compile_jar,
-                        java_toolchain = toolchains.java,
-                    ),
-                ],
-            )
+            kt_compile_jar = kt_runtime_jar
+            outputs = {
+                "output": kt_runtime_jar,
+            }
 
-    _run_kt_builder_action(
-        ctx = ctx,
+        _run_kt_builder_action(
+            ctx = ctx,
+            rule_kind = rule_kind,
+            toolchains = toolchains,
+            dirs = dirs,
+            srcs = srcs,
+            generated_src_jars = generated_src_jars,
+            friend = friend,
+            compile_deps = compile_deps,
+            annotation_processors = [],
+            transitive_runtime_jars = transitive_runtime_jars,
+            plugins = plugins,
+            outputs = outputs,
+            build_java = False,
```

Comment:
I left these janky `build_java` & `build_kotlin` flags in here, but I'm hoping we can just remove the in-KotlinWorker Java compilation and simplify things.

Reply:
I think, if you want to keep the existing in-Kotlin worker Java compilation as the main code path (and this new stuff behind a flag) then probably I should keep the build_java / build_kotlin flags for clarity / simplicity? Otherwise we're doing a lot of if/elif checks on the various inputs to determine which mode to run the builder in.